src/maintenance.js

```diff
@@ -133,3 +134,167 @@ module.exports.rotate = async function({Namespace, now, cfg, rabbitManager}) {
+The number of messages queued in \`${queue.name}\` exceeded ${cfg.deleteThreshold}.
+At the time of deletion, there were ${queue.messages} messages in the queue.`;
+  } else if (queue.messages > cfg.alertThreshold) {
+    currentState = 'warning';
+    subject = `${queue.name} has is in danget of being deleted!`;
+    content = `
+The number of messages queued in \`${queue.name}\` is now above ${cfg.alertThreshold}.
+Currently there are ${queue.messages} messages in the queue. If this number goes
+above ${cfg.deleteThreshold}, the queue will be deleted and all of the messages
+will be lost.
+
+A common cause of this situation is that your service has crashed.`;
+  }
+  return {currentState, subject, content};
+}
+
+async function needMessage(name, currentState, RabbitQueue) {
+  let rq;
+  let sendMessage = false;
+  try {
+    rq = await RabbitQueue.load({name});
+
+    rq.modify((entity) => {
+      if (entity.state !== currentState) {
+        entity.state = currentState;
+        entity.updated = new Date();
+        sendMessage = true;
+      }
+    });
+  } catch (e) {
+    if (e.code !== 'ResourceNotFound') {
+      throw e;
+    }
+    rq = await RabbitQueue.create({
+      name,
+      state: currentState,
+      updated: new Date(),
+    });
+    if (currentState !== 'normal') {
+      sendMessage = true;
+    }
+  }
+  return sendMessage;
+}
+
+async function handleQueues({cfg, prefix, manager, Namespace, RabbitQueue, notify}) {
+  let debug = Debug('maintenance.handle-queues');
+  await Promise.map(await manager.queues(), async queue => {
+    if (!queue.name.startsWith(cfg.queuePrefix + prefix)) {
+      return; // Note: This is very important to avoid stepping on pulseguardian's toes
+    }
+
+    let {currentState, subject, content} = decideState(queue, cfg);
+    let namespace = queue.name.split('/')[1];
+    let ns = await Namespace.load({namespace}, true);
+
+    if (!ns) {
+      // We get rid of any queues from namespaces that are gone
+      return await manager.deleteQueue(queue.name);
+    }
+
+    // First we'll send any notifications that we can
+    if (await needMessage(queue.name, currentState, RabbitQueue)) {
+
+      if (ns.contact) {
+        debug(`Sending a ${currentState} notification for ${queue.name} to ${ns.contact}`);
+        await notify.email({
+          address: ns.contact,
+          subject,
+          content,
+        });
+      } else {
+        debug(`Skipped sending a notification for ${queue.name} because no contact specified`);
+      }
+    }
+
+    // Finally we'll delete the queues if they're danger-big
+    if (currentState === 'danger') {
+      debug(`deleting ${queue.name} with ${queue.messages} messages.`);
+      await manager.deleteQueue(queue.name);
+    }
+  }, {concurrency: 10});
+}
+
+async function handleExchanges({cfg, prefix, manager, Namespace}) {
+  let debug = Debug('maintenance.handle-exchanges');
+  await Promise.map(await manager.exchanges(), async exchange => {
```

Comment:
This will end up calling `Namespace.load` many times with the same argument, in a case where a namespace has a lot of exchanges, right?  It might be nice to pre-flight that, parsing the exchange list into {namespace, exchanges: []..}.  Similarly for queues.  In practice I don't know if this is a big deal -- most users probably only have <10 exchanges and queues..